### PR TITLE
feat: init ring implement.

### PR DESCRIPTION
**Pull request by IRONICBo**
Implement a simpe consistant hash ring by `range` array, support balance and expand.

The slot struct contains `start` and `end` range, a hashed key can be mapped to a index in slot.
```rs
pub struct Slot<T> 
where T: NodeType
{
    /// The start offset of the slot
    start: u64,
    /// The end offset of the slot
    end: u64,
    /// The slot data, contains mapping info
    inner: T,
}
```

The `Ring` struct contains a hash method, slots, capacity(init with 1024), version(store add/remove version), this struct is not concurrent security, we need to use `Mutex` to share ring state.
```rs
#[derive(Debug, Clone)]
#[allow(dead_code)]
pub struct Ring<T, S = DefaultHashBuilder>
where T: NodeType,
      S: BuildHasher
{
    /// The hash builder
    hash_builder: S,
    /// The slots
    slots: Vec<Slot<T>>,
    /// T to slot id mapping, accelerate finding the slot
    /// The slot step of the ring
    capacity: u64,
    /// The version of the ring
    version: u64,
}
```

### CODE REVIEW

**Comment by frostyplanet** [comment]
pub fn get_nodes() 函数没看懂， 什么情况下会一个 key 对应多个 node? 用在什么地方

File: src/storage/cache_proxy/ring.rs [hunk 1]
```diff
@@ -0,0 +1,790 @@
...
+
+            return Some(node.clone());
+        }
+
+        // Try to judge if the ring need to be expanded
+        let thres = match self.capacity.checked_mul(RING_LOAD) {
+            Some(thres) => thres.overflowing_div(100).0,
+            None => return None,
+        };
+        while usize_to_u64(self.slots.len()) >= thres && must {
```
Comment (frostyplanet) [inline src/storage/cache_proxy/ring.rs:222]: 这个地方不是循环没必要用 while吧？

File: src/storage/cache_proxy/ring.rs [hunk 2]
```diff
@@ -0,0 +1,790 @@
...
+
+        // Update the end of the existing slot to the mid_point
+        slot_to_split.end = mid_point;
+
+        // Insert the new slot to index+1, and shift the rest of the slots
+        self.slots.insert(index + 1, new_slot);
+
+        // Try to rebalance the ring
+        // If must is true and the rebalance failed, return None
+        if must && !self.rebalance() {
```
Comment (frostyplanet) [inline src/storage/cache_proxy/ring.rs:254]: rebalance 应该不存在 fail 的情况， 既然节点已经加进去了, 就肯定不是 empty

**Comment by frostyplanet** [comment]
由于 meta client 的文件现在还没实现，我觉得 MetaClient 这个  trait 的接口定义得不是很实用.
应该只需要定以下几个接口：
* register()  节点上线注册
* watch() 等待节点上下线事件的通知
* save_ring() 保存
* load_ring() 加载

**Comment by frostyplanet** [comment]
1. 目前应该没有 replicas 的需求，所以这部分函数应该不用实现
2. pub fn get_nodes() 函数没看懂， 什么情况下会一个 key 对应多个 node? 用在什么地方
3. 对于 Node, id 有什么意义，id 由谁分配？如何取值
4. 灵魂问题：slot 有什么存在的必要。
现在看这个实现 rebalance 是给各个node 等分， slot 和 node 是一一对应的关系，那么 就没有 slot这个概念存在的必要。 
假如是用一致性 hash的做法， 那么 一个 node 就可能对应 多个 slot(vnode)。
初始化 capacity 默认 1024也没必要, 直接把 capacity 等于 u64 最大值 就不需要 expand了。

**Review by frostyplanet — CHANGES_REQUESTED** [review]

File: Cargo.toml [hunk 3]
```diff
@@ -68,6 +68,7 @@ opendal = {version = "0.43.0", features = ["layers-prometheus"]}
 signal-hook = { version = "0.3.17", features = ["iterator"]}
 signal-hook-tokio = { version = "0.3.1", features = ["futures-v0_3"] }
 tokio-util = "0.7.10"
+siphasher = "0.3.1"
```
Comment (frostyplanet) [inline Cargo.toml:71]: Better not introduce new dependency, unless it's a "must".
There are too many hasher out there, need throughout investigate to pick one

**Review by frostyplanet — CHANGES_REQUESTED** [review]

**Comment by frostyplanet** [comment]
the change in the latest commit "[feat: Update meta client.](https://github.com/datenlord/datenlord/pull/506/commits/cc589d45f1d7ef772f08164e9fba439a47c90cc8)"
should be split apart.  

Those fix the previous commit better be merged with the previous commits.

Unfinished commented code  better not submit now,  or mark the commit as "temp" to indicate it will be rewrite later

**Comment by IRONICBo** [comment]
> the change in the latest commit "[feat: Update meta client.](https://github.com/datenlord/datenlord/pull/506/commits/cc589d45f1d7ef772f08164e9fba439a47c90cc8)" should be split apart.
> 
> Those fix the previous commit better be merged with the previous commits.
> 
> Unfinished commented code better not submit now, or mark the commit as "temp" to indicate it will be rewrite later

Etcd part has moved to https://github.com/datenlord/datenlord/pull/514

**Pull request merged**